### PR TITLE
Add anchor to output of getblock

### LINKED ISSUE #2164: add anchor to output of getblock
Add the note commitment anchor to the JSON output.

### PR DESCRIPTION
**Pull request by str4d**
Closes #2164.

### CODE REVIEW

**Comment by bitcartel** [comment]
cc @daira 

Can we keep it simple and just return the hashAnchor which we already have in CBlockIndex?  It's a valid treestate which we can use for the purposes of anchoring right?

Also why give the user the end tree state as opposed to the start tree state for the block (which is the end tree state of the previous block)?  Or give them both - but that would be confusing.

**Comment by str4d** [comment]
The tree state that we currently anchor on for transactions is the state at the chain tip, ie at the end of the current best block. As I understand the protocol spec, the anchor for a block is supposed to be after including all of the JoinSplits within that block.

I added `hashAnchor` when implementing witness caching. I needed to get the tree state at the start of the block, so I could increment from there and pick up witnesses partway within a block. If I had understood the Bitcoin Core index architecture better at the time, I would have made `hashAnchor` be for the tree state at the end of the block, and used `pindex->pprev->hashAnchor` in `IncrementNoteWitnesses`. Then, this PR would simply have exposed `hashAnchor`. As-is, we either need to do what I'm doing here (cache `hashAnchorEnd` in-memory), or write migration code to re-define `hashAnchor` as the anchor for the end of a block (which would still mean having this code for older blocks, but simpler code for newer blocks or after a re-index). I don't know what the migration strategy would be for that however, and as it would still involve this code, this PR seems like the simplest way forward for now.

**Comment by daira** [comment]
I don't think that having "anchor_before" and "anchor_after"  keys would be confusing.

**Comment by str4d** [comment]
It may not be confusing, but we still want to ensure that the user would sensibly use what we provide. Given that the spec states that (ignoring interstitials) a JoinSplit **MUST** anchor on the output tree state of a block, the anchor that any transaction will use **MUST** be at the end of a block. Thus I don't see the point of including `anchor_before`, given that the user can a) just call `getblock` on the current block's parent, and b) is (I assume) wanting to know what the anchor for the *current* block is.

File: src/main.cpp [hunk 1]
```diff
@@ -3579,11 +3584,27 @@ bool static LoadBlockIndexDB()
...
+        }
+    }
+
     // Load pointer to end of best chain
     BlockMap::iterator it = mapBlockIndex.find(pcoinsTip->GetBestBlock());
     if (it == mapBlockIndex.end())
         return true;
     chainActive.SetTip(it->second);
+    // Set hashAnchorEnd for the end of best chain
+    it->second->hashAnchorEnd = pcoinsTip->GetBestAnchor();
```
Comment (bitcartel) [inline src/main.cpp]: Could this get out of sync?  That is, GetBestAnchor() doesn't return the anchor for the bestblock?

Comment (bitcartel) [inline src/main.cpp]: @str4d I just have this comment/question on this PR.

**Comment by bitcartel** [comment]
ACK (one comment/question above for @str4d, and also, do we need a test for this?)

**Comment by daira** [comment]
@str4d wrote:
> The tree state that we currently anchor on for transactions is the state at the chain tip, ie at the end of the current best block.

That's true, but we probably **shouldn't** be doing that (see #1614).

That said, I'm reasonably happy with the PR as-is. We discussed the motivation for exposing the final rather than initial treestate anchor in the pairing on Tuesday (basically, the final anchor is what would have been tracked if we'd thought about it more carefully in advance, and we may still change it to do that). We can always add the initial anchor to the output later.

**Comment by daira** [comment]
utACK. @zkbot r+

**Pull request merged**